wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -0,0 +1,147 @@
+## Table of Contents
```

Comment:
I think this is an excellent start. Some other things that I think are worth noting:
- We aim for 100% behavioral coverage, not just line of code coverage.
- The concept of a class's 'public API' (that is, its public members and how they behave in different contexts or with different call values/states).
- How to map a line of code to its corresponding behaviors (i.e. mapping it back to its public API and the behaviors that can lead to it being called). This is a _very_ useful thing to write up since it allows us to provide developers with a basic guide on how to effectively improve code coverage when writing tests.

Perhaps it's also worth framing the naming style that you repeat in this page in terms of behaviors: tests ought to verify a single behavior, and a behavior is:
- Tied to a single action (like a method call).
- Requires specific state to occur (its preconditions).
- Can be observed in a number of ways, but the test only needs to verify it in a single way (the expected outcome).

Reply:
Thank you so much for providing a detailed breakdown of the page requirements and specific sections. That was a beacon.

- I’ve added the behavioral section to address the behavioral coverage requirements, and the writing effective tests have been updated to align with these requirements - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#writing-effective-tests . 

- For the public API, I’ve included a sample general function to illustrate the scenario and demonstrate various ways to test and map it. However, I’m unsure about all possible states specific to the Oppia codebase. - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#guidelines-for-testing-public-api

- Added the "How to map a line of Code" section and with the mapping section, I used the same public API to explain a basic scenario. Should there be any additional complex scenarios that should be included, particularly those specific to the codebase? - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#how-to-map-a-line-of-code-to-its-corresponding-behaviors

- I’ve also added the name framing to the testing practices section - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#3-descriptive-test-names 

- Included tests to illustrate the need for specific conditions. -  https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#3-focusing-on-specific-test-cases

- Also, added the section to explain how a single outcome can be tested in different ways. - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#8-testing-a-single-outcome-in-multiple-ways

Reply:
Thanks! This is a really excellent set of changes and I've gone through, reviewed them, and commented.

Re: public API, I actually have a different perspective on how to think about this (behavior first rather than state based). See my newer comment on this specific concept.

Re: mapping to line of code, I actually think your example is really good. I don't see a need for a more complex example at this time.